pkg/agent/proxy/proxier.go

```diff
@@ -1179,18 +1181,34 @@ func (p *proxier) GetServiceByIP(serviceStr string) (k8sproxy.ServicePortName, b
 	return serviceInfo, exists
 }
 
-func (p *proxier) addServiceByIP(serviceStr string, servicePortName k8sproxy.ServicePortName) {
+// addServiceByIP adds ServicePortName entries to the proxier's serviceStringMap.
+//
+// The ServicePortName can be looked up by either the ClusterIP or the External
+// LoadBalancer IP.
+func (p *proxier) addServiceByIP(serviceInfo *types.ServiceInfo, servicePortName k8sproxy.ServicePortName) {
 	p.serviceStringMapMutex.Lock()
 	defer p.serviceStringMapMutex.Unlock()
 
+	serviceStr := serviceInfo.String()
 	p.serviceStringMap[serviceStr] = servicePortName
+
+	for _, serviceStr := range serviceInfo.GetLoadBalancerVIPStrings() {
+		p.serviceStringMap[serviceStr] = servicePortName
+	}
 }
 
-func (p *proxier) deleteServiceByIP(serviceStr string) {
+func (p *proxier) deleteServiceByIP(serviceInfo *types.ServiceInfo) {
+	p.deleteServiceIPs(
+		append([]string{serviceInfo.String()}, serviceInfo.GetLoadBalancerVIPStrings()...),
+	)
+}
+
+func (p *proxier) deleteServiceIPs(serviceStrings []string) {
```

Comment:
A method name like this feels a bit generic. More generally I think the names of the methods that manipulate `serviceStringMap` are not great.
Could we consider wrapping `serviceStringMap` and its mutex in a new struct? Something like:
```
type ipToServiceMap struct {
    mutex sync.Mutex
    m map[string]string
}
```

then we can move these methods to this new type.

Another advantage I can see is that it prevents the mutex (currently `p.serviceStringMapMutex`) from being misused for something else.

Reply:
I don't think we need his wrapper method?
same for `addServiceByIP` and `deleteServiceByIP`?